mnekrasov/files-file-systems/reverse-copying/main.c

```diff
@@ -0,0 +1,187 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <dirent.h>
+#include <sys/stat.h>
+#include <errno.h>
+
+#define SUCCESS 0
+#define ERROR -1
+#define EQUAL 0
+
+void reverseString(char *string) {
+    int stringLen = strlen(string);
+
+    for (int i = 0; i < stringLen / 2; ++i) {
+        char temp = string[i];
+        string[i] = string[stringLen - 1 - i];
+        string[stringLen - 1 - i] = temp;
+    }
+}
+
+void reverseFile(const char *srcFilePath, const char *dstFilePath) {
+    FILE *srcFile = fopen(srcFilePath, "rb");
+    if (!srcFile) {
+        perror("Impossible to open source file for reading");
+        return;
+    }
+
+    fseek(srcFile, 0, SEEK_END);
+    long srcFileSize = ftell(srcFile);
+
+    char *buffer = (char *) malloc(srcFileSize * sizeof(char));
+    if (buffer == NULL) {
+        perror("Impossible to allocate memory for buffer");
+        fclose(srcFile);
+        return;
+    }
+    fseek(srcFile, 0, SEEK_SET);
+    fread(buffer, sizeof(char), srcFileSize, srcFile);
+    fclose(srcFile);
+
+    FILE *dstFile = fopen(dstFilePath, "wb");
+    if (!dstFile) {
+        perror("Impossible to open destination file for writing");
+        free(buffer);
+        return;
+    }
+
+    for (long i = srcFileSize - 1; i >= 0; --i) {
+        fwrite(&buffer[i], sizeof(char), 1, dstFile);
+    }
+
+    fclose(dstFile);
```

Comment:
> Dealing with error returns from close()
A  careful  programmer will check the return value of close(), since it is quite possible that errors on a previous write(2) operation are reported only on the final close() that releases the open file description.  Failing to check the return value when closing a file may lead to silent loss of data.  This can especially be observed with NFS and with disk quota.